R/add_cols.R

```diff
@@ -0,0 +1,247 @@
+#' Add event date as column to infectious history `<data.frame>`
+#'
+#' @description The event date could be first contact, last contact or other.
+#'
+#' @param .data A `<data.frame>` containing the infectious history from a
+#' branching process simulation.
+#' @param contact_type A `character` with the type of contact, either first
+#' contact (`"first"`), or last contact (`"last"`).
+#' @param distribution A `character` with the name of the distribution,
+#' following the base R convention for distribution naming (e.g. Poisson
+#' is `pois`).
+#' @param ... [dots] Extra arguments to be passed to the distribution function
+#' given in the `distribution` argument.
+#' @inheritParams sim_linelist
+#'
+#' @name .add_date
+#'
+#' @return A `<data.frame>` with one more column than input into `.data`.
+#' Unless the column heading is already present in which the data is
+#' overwritten.
+#' @keywords internal
+#'
+NULL
+
+#' @name .add_date
+.add_date_contact <- function(.data,
+                              contact_type = c("first", "last"),
+                              distribution = c("pois", "geom"),
+                              ...,
+                              outbreak_start_date) {
+  contact_type <- match.arg(contact_type)
+  distribution <- match.arg(distribution)
+
+  stopifnot(
+    "outbreak_start_date is only required for adding date of last contact" =
+      contact_type == "last" && !missing(outbreak_start_date) ||
+      contact_type == "first" && missing(outbreak_start_date)
+  )
+
+  rdist <- switch(distribution,
+    pois = stats::rpois,
+    geom = stats::rgeom
+  )
+
+  # c() over ...length() to ensure NULL is not counted by length
+  if (length(c(...)) == 0) {
+    stop("Distribution parameters are missing, check config", call. = FALSE)
+  }
+
+  # name list elements with vec names to ensure arg matching in do.call
+  args <- c(nrow(.data), list(...))
+
+  contact_delay <- tryCatch(
+    error = function(cnd) {
+      stop(
+        "Incorrect parameterisation of distribution, check config",
+        call. = FALSE
+      )
+    },
+    warning = function(cnd) {
+      stop(
+        "Incorrect parameterisation of distribution, check config",
+        call. = FALSE
+      )
+    },
+    do.call(rdist, args = args)
+  )
+
+  if (contact_type == "first") {
+    .data$date_first_contact <- .data$date_last_contact - contact_delay
+  } else {
+    .data$date_last_contact <- .data$infector_time + contact_delay +
+      outbreak_start_date
+  }
+
+  # return data
+  .data
+}
+
+#' @name .add_date
+.add_hospitalisation <- function(.data,
+                                 onset_to_hosp,
+                                 hosp_risk) {
+  infected_idx <- .data$infected == "infected"
```

Comment:
I know this is an internal method and want to check-in: How robust is this implementation of identifying the infected in the data?

Because we don't have [strong types](https://en.wikipedia.org/wiki/Strong_and_weak_typing), I wonder how certain we can be that the infected are identified in exactly this way (and not for example "Infected"). If we can be certain, then this is 👍  - if we remain uncertain I wonder how we might improve this.

Reply:
I found where it comes from in `R/sim_network_bp.R` line 113 - that means it is implicitly typed as such. Good to make a note over there that if it is changed all the booleans need to be changed too (I'll suggest).

Reply:
Thanks for pointing this out. It's definitely not the safest solution. I'll add your suggested comment below. I can also change the implementation if you think of a more secure way to define and check against this variable.